src/core/execute.c

```diff
@@ -361,9 +360,13 @@ static int connect_logger_as(
                 !!context->syslog_level_prefix,
                 is_syslog_output(output),
                 is_kmsg_output(output),
-                is_terminal_output(output));
+                is_terminal_output(output)) < 0)
+                return -errno;
 
-        return move_fd(fd, nfd, false);
+        if (dprintf(fd, " ") < 0)
+                return -errno;
```

Comment:
why this bit?

Reply:
For some reason that I haven't fully figured out yet, sending this "canary" is the only reliable way to trigger `EPIPE`.   I find it a bit kludgy it so I added the "don't merge" label.

Reply:
that sounds like a race to me: given that the other side has to accept the connections before it can close them (and inbetween any time can pass) it might very well happen that we don't see the disconnect on the pid1 side before we hand off the fd to the activated program...

Reply:
Like @poettering said there's a window between the `accept4()` and the `close()` in journald when the max is exceeded where PID1 can be scheduled and proceeds as if everything's fine.

Without adding more explicit coordination between journald and PID1 surrounding the handoff, I think the only robust alternative is when the maximum has been reached in journald instead of accepting the connection at all, close the listener which IIRC fails the connect, and all subsequent connects will fail.  When the # of connected streams drops below the maximum, then restore the listener.

Reply:
`systemd` itself is a listener (if `systemd-journald.socket` is active) so it seems that `connect` will succeed as long as `systemd` keeps listening. Apparently, at some point `systemd` could even block itself waiting for `connect`.

Reply:
Though, in terms of user-friendliness, getting `ECONNREFUSED` is much better than hoping that `SIGHUP` won't be received after sending a message :-) On the other hand, closing the listening side doesn't seem to help when the first client is refused. It will get `EPIPE` and `SIGPIPE` as usual. I'm not sure it can be fixed without more explicit coordination.

Reply:
It gets even more annoying; the way AF_UNIX is currently implemented in the kernel the connects are all successful immediately (as in, no EINPROGRESS) within the bounds of the listen backlog+1.

Even if you set the backlog to 0, one queued connect will immediately succeed independent of a corresponding `accept()`.  So there's no way to have an AF_UNIX listener which blocks *any* connects from occurring until calling `accept()`.  I think this is a bug; it seems one should be able to make `connect()` synchronous with `accept()` on a given listener by setting the backlog to 0.

As-is, even if we managed to close the listener when the max was reached, there'd still be a small race where the kernel could have queued another successful connect on the listener between our limit-reaching `accept()` and the `close()` of the listener.